### PR TITLE
Implement RowReader::Cancel()

### LINKED ISSUE #179: Implement bigtable::RowReader::Cancel()
This function is declared, but not implemented or tested.  At the end of this task the function should exist and be able to cancel a stream safely.  Please note that gRPC requires one to read all the elements of the stream before calling `Finish()`.

### PR DESCRIPTION
**Pull request by DanielMahu**
Fixes #179.

### CODE REVIEW

File: bigtable/client/row_reader_test.cc [hunk 1]
```diff
@@ -218,6 +218,32 @@ TEST_F(RowReaderTest, ReadOneRowIteratorPostincrement) {
...
+  bigtable::RowReader reader(
+      client_, "", bigtable::RowSet(), bigtable::RowReader::NO_ROWS_LIMIT,
+      bigtable::Filter::PassAllFilter(), std::move(retry_policy_),
+      std::move(backoff_policy_), std::move(parser_factory_));
+
+  auto it = reader.begin();
+  EXPECT_NE(it, reader.end());
+  EXPECT_EQ(it->row_key(), "r1");
+  EXPECT_NE(it, reader.end());
+  // Do not finish iterating, and expect the stream to be finalized anyway.
```
Comment (coryan) [inline bigtable/client/row_reader_test.cc]: I think you are relying on the expectations setup above in this test.  Can you clarify the comment to say that?  For example:

```C++
// Do not finish the iteration.  We expect the stream to be finalized anyway.  The previously
// setup expectations on the mock `stream_` check that.
```

File: bigtable/client/row_reader_test.cc [hunk 2]
```diff
@@ -525,3 +551,42 @@ TEST_F(RowReaderTest, RowLimitIsNotDecreasedToZero) {
...
+      bigtable::Filter::PassAllFilter(), std::move(retry_policy_),
+      std::move(backoff_policy_), std::move(parser_factory_));
+
+  auto it = reader.begin();
+  EXPECT_NE(it, reader.end());
+  EXPECT_EQ(it->row_key(), "r1");
+  EXPECT_NE(it, reader.end());
+  // Manually cancel the call.
+  reader.Cancel();
+  EXPECT_THROW(reader.begin(), std::runtime_error);
```
Comment (coryan) [inline bigtable/client/row_reader_test.cc]: In general, what should be the effect of calling `begin()` twice?  No action on this PR, just wondering.

File: bigtable/client/row_reader_test.cc [hunk 3]
```diff
@@ -525,3 +551,42 @@ TEST_F(RowReaderTest, RowLimitIsNotDecreasedToZero) {
...
+}
+
+TEST_F(RowReaderTest, BeginThrowsAfterImmediateCancel) {
+  bigtable::RowReader reader(
+      client_, "", bigtable::RowSet(), bigtable::RowReader::NO_ROWS_LIMIT,
+      bigtable::Filter::PassAllFilter(), std::move(retry_policy_),
+      std::move(backoff_policy_), std::move(parser_factory_));
+  // Manually cancel the call before a stream was created.
+  reader.Cancel();
```
Comment (coryan) [inline bigtable/client/row_reader_test.cc]: Do we have expectations that the reader will **not** do things in this case?  For example, would it be bad if it issued RPCs?  If so, we should have a few expectations with `.Times(0)` or something.

File: bigtable/client/row_reader_test.cc (hunk 1 above)
Comment (DanielMahu) [reply bigtable/client/row_reader_test.cc]: done

File: bigtable/client/row_reader_test.cc (hunk 2 above)
Comment (DanielMahu) [reply bigtable/client/row_reader_test.cc]: I think it's OK to allow it to continue reading the rows not touched by the other iterator, and even allow the two to advance in arbitrary order, but I can't find something similar and documented to do this. I can almost hear a whisper: undefined...

File: bigtable/client/row_reader_test.cc (hunk 3 above)
Comment (DanielMahu) [reply bigtable/client/row_reader_test.cc]: done

**Comment by coryan** [comment]
SGTM.  Note that the Travis build is failing because of the problem identified in #187.

**Comment by coryan** [comment]
@DanielMahu I think you will need to rebase and push again to get Travis to compile this correctly.  Do you mind doing so?

**Comment by coryan** [comment]
And now the Travis build has detected a memory leak:

https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/jobs/330433900#L9204

@DanielMahu in case you missed it.  This could be a leak in the test, not the code, but still needs fixing.

**Comment by DanielMahu** [comment]
Thanks! That's true, the test allocates unsafely the value returned by ReadRowsRaw.

**Pull request merged**